docs/annotated_mnist.md

```diff
@@ -1,157 +1,317 @@
 
-```py
-import tensorflow as tf
-import tensorflow_datasets as tfds
+
+```python
+def cross_entropy_loss(logits, labels):
+  one_hot_labels = jax.nn.one_hot(labels, num_classes=10)
+  return -jnp.mean(jnp.sum(one_hot_labels * logits, axis=-1))
 ```
 
-A Flax "module" lets you write a normal function, which
-defines learnable parameters in-line. In this case,
-we define a simple convolutional neural network.
+2. Define a function for your optimizer that takes the model parameters, the learning rate, and a beta argument (for the [Momentum optimizer](http://www.columbia.edu/~nq6/publications/momentum.pdf)):
 
-Each call to `flax.nn.Conv` defines a learnable kernel.
+  - Choose `Momentum` from the [`flax.optim`](https://flax.readthedocs.io/en/latest/flax.optim.html) package; and
+  - Wrap the model parameters (`params`) with the [`flax.optim.OptimizerDef.create`](https://flax.readthedocs.io/en/latest/flax.optim.html#flax.optim.OptimizerDef.create) method and initialize with parameter dicts.
 
-```py
-class CNN(flax.nn.Module):
-  def apply(self, x):
-    x = flax.nn.Conv(x, features=32, kernel_size=(3, 3))
-    x = flax.nn.relu(x)
-    x = flax.nn.avg_pool(x, window_shape=(2, 2), strides=(2, 2))
-    x = flax.nn.Conv(x, features=64, kernel_size=(3, 3))
-    x = flax.nn.relu(x)
-    x = flax.nn.avg_pool(x, window_shape=(2, 2), strides=(2, 2))
-    x = x.reshape((x.shape[0], -1))
-    x = flax.nn.Dense(x, features=256)
-    x = flax.nn.relu(x)
-    x = flax.nn.Dense(x, features=10)
-    x = flax.nn.log_softmax(x)
-    return x
+
+```python
+def create_optimizer(params, learning_rate, beta):
+  optimizer_def = optim.Momentum(learning_rate=learning_rate, beta=beta)
+  optimizer = optimizer_def.create(params)
+  return optimizer
 ```
 
-`jax.vmap` allows us to define the `cross_entropy_loss`
-function as if it acts on a single sample. `jax.vmap`
-automatically vectorizes code efficiently to run on entire
-batches.
+3. Create a function for parameter initialization:
 
-```py
-@jax.vmap
-def cross_entropy_loss(logits, label):
-  return -logits[label]
+  - Set the initial shape of the kernel (note that JAX and Flax are [row-based](https://flax.readthedocs.io/en/latest/notebooks/flax_basics.html#Model-parameters-&-initialization)); and
+  - Initialize the module parameters of your network (`CNN`) with the [`flax.linen.Module.init`](https://flax.readthedocs.io/en/latest/flax.linen.html#flax.linen.Module.init) method using the PRNGKey and parameters (note that the parameters aren't stored with the models).
+
+
+```python
+def get_initial_params(key):
+  init_shape = jnp.ones((1, 28, 28, 1), jnp.float32)
```

Comment:
This is an incorrect variable name -- this isn't a shape, it's an array.

Reply:
OK! 👍 The code came from our Flax `linen_examples` - maybe we should also update the MNIST example there. WDYT?

https://github.com/google/flax/blob/9fdf59f05702c746ea9a6be50cd3a81f38e2fa0b/linen_examples/mnist/train.py#L54

```py
def get_initial_params(key):
  init_shape = jnp.ones((1, 28, 28, 1), jnp.float32)
  initial_params = CNN().init(key, init_shape)['params']
  return initial_params
```